server/tests/lib/topic_cache_test.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import os
+import shutil
+import tempfile
+import unittest
+from unittest.mock import patch
+
+from server.lib import topic_cache
+from server.lib.nl.explore.params import DCNames
+
+_TEST_MODE = os.environ.get('TEST_MODE', '')
+
+_TEST_DATA_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                              "test_data", "topic_cache")
+_INPUT_DIR = os.path.join(_TEST_DATA_DIR, "input")
+_EXPECTED_DIR = os.path.join(_TEST_DATA_DIR, "expected")
+_MAIN_DC_TOPIC_CACHE_FILE = os.path.join(_INPUT_DIR, "main_dc_topic_cache.json")
+_UNDATA_TOPIC_CACHE_FILE = os.path.join(_INPUT_DIR, "undata_topic_cache.json")
+# The custom dc topic cache json will be in a file
+# (datacommons/nl/custom_dc_topic_cache.json) relative to this path.
+_CUSTOM_DC_MERGE_WITH_MAIN_USER_DATA_PATH = os.path.join(
+    _INPUT_DIR, "merge_with_main")
+_CUSTOM_DC_MERGE_WITH_UNDATA_USER_DATA_PATH = os.path.join(
+    _INPUT_DIR, "merge_with_undata")
+
+
+def _compare_files(test: unittest.TestCase, output_path: str,
+                   expected_path: str, message: str):
+  with open(output_path) as gotf:
+    got = gotf.read()
+    with open(expected_path) as wantf:
+      want = wantf.read()
+      test.assertEqual(got, want, message)
+
+
+def _test_topic_cache_loader(test: unittest.TestCase, test_name: str):
+  test.maxDiff = None
+
+  with tempfile.TemporaryDirectory() as temp_dir:
+    output_json_path = os.path.join(temp_dir, f"{test_name}.json")
+    expected_json_path = os.path.join(_EXPECTED_DIR, f"{test_name}.json")
+
+    topic_caches = topic_cache.load({})
+    # Write json
+    topic_caches_json = {
+        dc: topic_cache.json() for dc, topic_cache in topic_caches.items()
+    }
+    with open(output_json_path, "w") as out:
+      json.dump(topic_caches_json, out, indent=1)
+
+    if _TEST_MODE == "write":
+      shutil.copy(output_json_path, expected_json_path)
+      return
+
+    _compare_files(test, output_json_path, expected_json_path,
+                   f"Found diffs in topic caches JSON: {test_name}")
+
+
+@patch.dict(topic_cache.TOPIC_CACHE_FILES, {
+    DCNames.MAIN_DC.value: [_MAIN_DC_TOPIC_CACHE_FILE],
+    DCNames.UNDATA_DC.value: [_UNDATA_TOPIC_CACHE_FILE]
+},
+            clear=True)
+class TestTopicCacheLoader(unittest.TestCase):
+
+  @patch.dict(os.environ, {"IS_CUSTOM_DC": "false"})
```

Comment:
Appreciate the detailed tests here, but could you please add some comments on what we're doing?

Reply:
Done.